homeassistant/components/vesync/sensor.py

```diff
@@ -21,32 +21,29 @@
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up switches."""
 
-    async def async_discover(devices):
+    @callback
+    def discover(devices):
         """Add new devices to platform."""
-        _async_setup_entities(devices, async_add_entities)
+        _setup_entities(devices, async_add_entities)
 
-    disp = async_dispatcher_connect(
-        hass, VS_DISCOVERY.format(VS_SENSORS), async_discover
-    )
+    disp = async_dispatcher_connect(hass, VS_DISCOVERY.format(VS_SENSORS), discover)
```

Comment:
To register the unsubscribe function to be called when unloading the config entry we can pass the unsubscribe function to`config_entry.async_on_unload`. Then we don't need to store the unsubscribe function in hass.data.


Here's how we do it in zwave_js:
https://github.com/home-assistant/core/blob/d0578509710ac27a20d1e5576e7eee15d140057d/homeassistant/components/zwave_js/sensor.py#L233-L239

Reply:
Thank you for the pointer. I've applied this change to all the platforms in this component in the latest commit. Also applied your other suggestions to the other components.